aten/src/ATen/native/cuda/Copy.cu

```diff
@@ -174,13 +174,17 @@ static void copy_kernel_cuda(TensorIterator& iter, bool non_blocking) {
   int64_t nbytes = iter.numel() * iter.element_size(0);
   CUDAStream stream = getCurrentCUDAStream();
 
-  AT_CUDA_CHECK(cudaMemcpyAsync(dst, src, nbytes, kind, stream));
-
   if (non_blocking) {
+    AT_CUDA_CHECK(cudaMemcpyAsync(dst, src, nbytes, kind, stream));
     void* ptr = (dst_device == kCPU ? dst : src);
     AT_CUDA_CHECK(THCCachingHostAllocator_recordEvent(ptr, stream));
   } else {
+#if HIP_VERSION >= 310
+    AT_CUDA_CHECK(hipMemcpyWithStream(dst, src, nbytes, kind, stream));
+#else
```

Comment:
Just want to make sure I understand this new API -- hipMemcpyWithStream waits for the entire copy to finish on a stream, is that right? So if I have multi-threads launching the copy, the copy can be in parallel -- but the API will hold each thread until it finishes?

Reply:
@xw285cornell concerning hipMemcpyWithStream, it directly replaces, optimizes, and behaves as if you'd write hipMemcpyAsync(stream) && hipStreamSynchronize(stream). So serialization is local to the stream (of course null stream has special properties here). So your statement is correct, multiple streams can be launched to and copy in parallel, but the API will be synchronous in that stream.

Reply:
I see sounds good. Just curious, since you mentioned this API will bring perf gains, just curious, what kind of optimizations happen under the hood for this optimization? 

The use case I was thinking about was unpinned memory, but in a different scenario. If you stay within pytorch, all cpu memory should be pinned.